### PR TITLE
Replay lost `MonitorEvent`s in some cases for closed channels

### PR DESCRIPTION
**Pull request by TheBlueMatt**
This is the first few commits from #3984 which handles some of the trivial cases, but skips actually marking `MonitorEvent`s completed which is needed for some further edge-cases.

### CODE REVIEW

**Comment by valentinewallace** [comment]
One of the new tests is failing

**Comment by TheBlueMatt** [comment]
Oops, silent rebase conflict due to #4001.

**Comment by valentinewallace** [comment]
Test is still failing

File: lightning/src/ln/channelmanager.rs [hunk 1]
```diff
@@ -16489,6 +16489,10 @@ where
 			// payments which are still in-flight via their on-chain state.
 			// We only rebuild the pending payments map if we were most recently serialized by
 			// 0.0.102+
+			//
+			// First we rebuild the pending payments, and only once we do so we go through and
+			// re-claim and re-fail pending payments. This avoids edge-cases around MPP payments
+			// resulting in redundant actions.
```
Comment (valentinewallace) [inline lightning/src/ln/channelmanager.rs]: I think the commit message should say s/PaymentClaimed/Payment*Sent*, since we're talking about outbound payments?

**Comment by TheBlueMatt** [comment]
Grr, the #4001 changes were random so had to fix all the paths, should be fine now.

File: lightning/src/ln/channelmanager.rs [hunk 2]
```diff
@@ -16527,6 +16531,18 @@ where
...
+				}
+			}
+			for (channel_id, monitor) in args.channel_monitors.iter() {
+				let mut is_channel_closed = false;
+				let counterparty_node_id = monitor.get_counterparty_node_id();
+				if let Some(peer_state_mtx) = per_peer_state.get(&counterparty_node_id) {
+					let mut peer_state_lock = peer_state_mtx.lock().unwrap();
+					let peer_state = &mut *peer_state_lock;
+					is_channel_closed = !peer_state.channel_by_id.contains_key(channel_id);
+				}
```
Comment (valentinewallace) [inline lightning/src/ln/channelmanager.rs]: I'm not seeing the bug that this is fixing -- if a payment is failed after being fulfilled in `outbound_payments`, we'll terminate early after noticing it's fulfilled already in `OutboundPayments::fail_htlc` 

Context from the commit message: "... we could get both a `PaymentFailed` and a `PaymentClaimed` event on startup for the same payment"

Comment (valentinewallace) [inline lightning/src/ln/channelmanager.rs]: Would appreciate some more context on this part of the commit message: "this can lead to a pending payment getting re-added and re-claimed multiple times" (phrased as bad). 

It looks like in the prior code and this new code, we'll call `OutboundPayments::insert_from_monitor_on_startup` for each `session_priv` (seems fine), and then call `claim_htlc` for each `session_priv` (seems fine since this will only generate a `PaymentSent` event on the first call to `claim_htlc`). Can you help me understand what I'm missing that was buggy in the prior code?

File: lightning/src/ln/monitor_tests.rs [hunk 3]
```diff
@@ -3290,3 +3291,167 @@ fn test_claim_event_never_handled() {
...
+			},
+			Event::PaymentForwarded { claim_from_onchain_tx, .. } => {
+				assert!(claim_from_onchain_tx);
+			},
+			_ => panic!("Wrong event {ev:?}"),
+		}
+	}
+
+	// After the background events are processed in `get_and_clear_pending_events`, above, node B
+	// will create the requisite `ChannelMontiorUpdate` for claiming the forwarded payment back.
```
Comment (valentinewallace) [inline lightning/src/ln/monitor_tests.rs:3451]: nit: ChannelMonitorUpdate

File: lightning/src/ln/monitor_tests.rs [hunk 4]
```diff
@@ -3290,3 +3291,167 @@ fn test_claim_event_never_handled() {
 	// `ChannelMonitorUpdate`.
 	check_added_monitors(&nodes[1], 2);
 }
+
+fn do_test_lost_preimage_monitor_events(on_counterparty_tx: bool) {
+	// `MonitorEvent`s aren't delivered to the `ChannelManager` in a durable fasion - if the
+	// `ChannelManager` fetches the pending `MonitorEvent`s, then the `ChannelMonitor` gets
+	// persisted (i.e. due to a block update) then the node crashes, prior to persisting the
+	// `ChannelManager` again, the `MonitorEvent` and its effects on the `ChannelManger` will be
```
Comment (valentinewallace) [inline lightning/src/ln/monitor_tests.rs:3309]: nit: ChannelManager

File: lightning/src/chain/channelmonitor.rs [hunk 5]
```diff
@@ -2957,6 +2957,75 @@ impl<Signer: EcdsaChannelSigner> ChannelMonitor<Signer> {
...
+		if confirmed_txid.is_none() {
+			return new_hash_map();
+		}
+
+		let mut res = new_hash_map();
+		macro_rules! walk_htlcs {
+			($holder_commitment: expr, $htlc_iter: expr) => {
+				for (htlc, source) in $htlc_iter {
+					let filter = |v: &&IrrevocablyResolvedHTLC| {
+						v.commitment_tx_output_idx == htlc.transaction_output_index
```
Comment (valentinewallace) [inline lightning/src/chain/channelmonitor.rs]: The output index can be `None` on both sides currently, is that intentional?

File: lightning/src/chain/channelmonitor.rs [hunk 6]
```diff
@@ -2957,6 +2957,75 @@ impl<Signer: EcdsaChannelSigner> ChannelMonitor<Signer> {
...
+					};
+					if let Some(state) = us.htlcs_resolved_on_chain.iter().filter(filter).next() {
+						if let Some(source) = source {
+							if state.payment_preimage.is_none() {
```
Comment (valentinewallace) [inline lightning/src/chain/channelmonitor.rs]: Tests pass when removing this check atm. Probably fine if the test is too annoying to write though.

File: lightning/src/chain/channelmonitor.rs [hunk 7]
```diff
@@ -2957,6 +2957,75 @@ impl<Signer: EcdsaChannelSigner> ChannelMonitor<Signer> {
 		res
 	}
 
+	/// Gets the set of outbound HTLCs which hit the chain and ultimately were claimed by us via
```
Comment (valentinewallace) [inline lightning/src/chain/channelmonitor.rs:3037]: Might be a dumb question, but can you clarify where we filter for outbound HTLCs in this method (i.e. ignore inbound ones)?

File: lightning/src/ln/channelmanager.rs (hunk 2 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs]: Lets say we send an MPP payment along two different channels. One is claimed and the other failed (who knows why, the recipient just decided they don't like money for whatever reason).

Going through the single loop we may first find the failed-htlc channel - we'll add the pending payment in `insert_from_monitor_on_startup` which will just add the one part, then we'll see it failed and generate a `PaymentFailed` event since there's only one part. Then we'll go to the second channel and repeat the same process, but now with a `PaymentSent` event.

Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs]: I rewrote the commit message.

File: lightning/src/chain/channelmonitor.rs (hunk 5 above)
Comment (TheBlueMatt) [reply lightning/src/chain/channelmonitor.rs]: Uhhh, it wasn't, but I believe it is almost the correct behavior - if we have a dust HTLC that never made it to chain and the commitment is confirmed, we want to consider it resolved. I'll tweak it to make it clearer and more likely correct.

File: lightning/src/chain/channelmonitor.rs (hunk 7 above)
Comment (TheBlueMatt) [reply lightning/src/chain/channelmonitor.rs:3037]: We check if `source` is `Some` - we only have sources for outbound HTLCs

File: lightning/src/chain/channelmonitor.rs (hunk 6 above)
Comment (TheBlueMatt) [reply lightning/src/chain/channelmonitor.rs]: I believe its unobservable. There are several cases where we'd hit a theoretical `else` block, but the `ChannelManager` read logic actually replays preimages first, so we'd always actually replay the preimage first, then get here, spuriously include it, and not have anything to fail.

File: lightning/src/ln/monitor_tests.rs [hunk 8]
```diff
@@ -3455,3 +3455,184 @@ fn test_lost_preimage_monitor_events() {
...
+		assert_eq!(bs_htlc_timeouts.len(), 1);
+
+		// Now replay the timeouts on node B, which after 6 confirmations should fail the HTLCs via
+		// `MonitorUpdate`s
+		mine_transactions(&nodes[1], &bs_htlc_timeouts.iter().collect::<Vec<_>>());
+		connect_blocks(&nodes[1], ANTI_REORG_DELAY - 1);
+	}
+
+	// Now simulate a restart where the B<->C ChannelMonitor has been persisted (i.e. because we
+	// just processed a new block) but the ChannelManager was not. This should be exceedingly rare
```
Comment (valentinewallace) [inline lightning/src/ln/monitor_tests.rs:3655]: In an early commit we remove a similar comment because monitors aren't persisted on new blocks, may want to amend here too?

File: lightning/src/ln/monitor_tests.rs [hunk 9]
```diff
@@ -3455,3 +3455,184 @@ fn test_lost_preimage_monitor_events() {
...
+				assert_eq!(prev_channel_id, chan_a);
+			},
+			_ => panic!("Wrong event {ev:?}"),
+		}
+	}
+
+	nodes[0].node.peer_disconnected(nodes[1].node.get_our_node_id());
+
+	let mut reconnect_args = ReconnectArgs::new(&nodes[0], &nodes[1]);
+	reconnect_args.pending_cell_htlc_fails = (0, 0);
```
Comment (valentinewallace) [inline lightning/src/ln/monitor_tests.rs]: This can be omitted because it's the default already

File: lightning/src/ln/channelmanager.rs [hunk 10]
```diff
@@ -16640,6 +16644,20 @@ where
 							},
 						}
 					}
+					for (htlc_source, htlc) in monitor.get_onchain_failed_outbound_htlcs() {
+						log_info!(
+							args.logger,
+							"Failing HTLC with payment hash {} as it was resolved on-chain.",
+							htlc.payment_hash
+						);
+						failed_htlcs.push((
```
Comment (valentinewallace) [inline lightning/src/ln/channelmanager.rs:16614]: Will we generate a `PaymentFailed` and/or `HTLCHandlingFailed` event for these HTLCs on every restart until the monitor is removed? Could be worth noting in a comment if so

File: lightning/src/ln/monitor_tests.rs (hunk 8 above)
Comment (TheBlueMatt) [reply lightning/src/ln/monitor_tests.rs:3655]: We sometimes persist monitors in response to new blocks, just not always.

File: lightning/src/ln/channelmanager.rs (hunk 10 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:16614]: That's fixed in #3984 :)

File: lightning/src/chain/channelmonitor.rs [hunk 11]
```diff
@@ -2957,6 +2957,75 @@ impl<Signer: EcdsaChannelSigner> ChannelMonitor<Signer> {
...
+							}
+						}
+					}
+				}
+			};
+		}
+
+		let txid = confirmed_txid.unwrap();
+		if Some(txid) == us.funding.current_counterparty_commitment_txid
+			|| Some(txid) == us.funding.prev_counterparty_commitment_txid
```
Comment (wpaulino) [inline lightning/src/chain/channelmonitor.rs]: Why don't we bother with counterparty revoked commitments?

File: lightning/src/chain/channelmonitor.rs [hunk 12]
```diff
@@ -2957,6 +2957,75 @@ impl<Signer: EcdsaChannelSigner> ChannelMonitor<Signer> {
...
+								res.insert(source.clone(), htlc.clone());
+							}
+						}
+					}
+				}
+			};
+		}
+
+		let txid = confirmed_txid.unwrap();
+		if Some(txid) == us.funding.current_counterparty_commitment_txid
```
Comment (wpaulino) [inline lightning/src/chain/channelmonitor.rs]: Doesn't this need to consider the current/prev txids for the `pending_funding` scopes as well if we have `alternative_funding_confirmed` set?

File: lightning/src/ln/monitor_tests.rs [hunk 13]
```diff
@@ -3455,3 +3455,180 @@ fn test_lost_preimage_monitor_events() {
 	do_test_lost_preimage_monitor_events(true);
 	do_test_lost_preimage_monitor_events(false);
 }
+
+fn do_test_lost_timeout_monitor_events(on_counterparty_tx: bool) {
+	// `MonitorEvent`s aren't delivered to the `ChannelManager` in a durable fasion - if the
```
Comment (wpaulino) [inline lightning/src/ln/monitor_tests.rs]: fashion* same typo across commit messages and tests

File: lightning/src/chain/channelmonitor.rs (hunk 11 above)
Comment (TheBlueMatt) [reply lightning/src/chain/channelmonitor.rs]: Ugh, you're right, this is buggy, I rewrote it.

File: lightning/src/chain/channelmonitor.rs (hunk 12 above)
Comment (TheBlueMatt) [reply lightning/src/chain/channelmonitor.rs]: It does, when it was written that PR hadn't landed yet :)

**Comment by TheBlueMatt** [comment]
Rebased to update against the new funding logic in ChannelMonitor.

**Comment by wpaulino** [comment]
Feel free to squash, this changed enough I'll be doing another full pass

**Comment by TheBlueMatt** [comment]
Squashed without further changes, @wpaulino

**Comment by TheBlueMatt** [comment]
Rebased and pushed with the following diff to account for the new channel closure monitor updates:

```diff
diff --git a/lightning/src/ln/monitor_tests.rs b/lightning/src/ln/monitor_tests.rs
index 609b39d336..8ac995e9b3 100644
--- a/lightning/src/ln/monitor_tests.rs
+++ b/lightning/src/ln/monitor_tests.rs
@@ -3425,7 +3425,7 @@ fn do_test_lost_preimage_monitor_events(on_counterparty_tx: bool) {
 	// latest state.
 	let mon_events = nodes[1].chain_monitor.chain_monitor.release_pending_monitor_events();
 	assert_eq!(mon_events.len(), 1);
-	assert_eq!(mon_events[0].2.len(), 2);
+	assert_eq!(mon_events[0].2.len(), 3);

 	let node_ser = nodes[1].node.encode();
 	let mon_a_ser = get_monitor!(nodes[1], chan_a).encode();
@@ -3661,7 +3661,7 @@ fn do_test_lost_timeout_monitor_events(confirm_tx: CommitmentType, dust_htlcs: b
 	// latest state.
 	let mon_events = nodes[1].chain_monitor.chain_monitor.release_pending_monitor_events();
 	assert_eq!(mon_events.len(), 1);
-	assert_eq!(mon_events[0].2.len(), 2);
+	assert_eq!(mon_events[0].2.len(), 3);

 	let node_ser = nodes[1].node.encode();
 	let mon_a_ser = get_monitor!(nodes[1], chan_a).encode();
```

File: lightning/src/ln/channelmanager.rs [hunk 14]
```diff
@@ -16489,6 +16493,18 @@ where
 							);
 						}
 					}
+				}
+			}
+			for (channel_id, monitor) in args.channel_monitors.iter() {
+				let mut is_channel_closed = false;
```
Comment (wpaulino) [inline lightning/src/ln/channelmanager.rs]: Shouldn't this be `true` in case we don't have the `per_peer_state` around either?

File: lightning/src/ln/channelmanager.rs [hunk 15]
```diff
@@ -16451,6 +16451,10 @@ where
 			// payments which are still in-flight via their on-chain state.
 			// We only rebuild the pending payments map if we were most recently serialized by
 			// 0.0.102+
+			//
+			// First we rebuild the pending payments, and only once we do so we go through and
+			// re-claim and re-fail pending payments. This avoids edge-cases around MPP payments
```
Comment (wpaulino) [inline lightning/src/ln/channelmanager.rs]: Nit: the first sentence here could be reworded a bit for brevity

File: lightning/src/chain/channelmonitor.rs [hunk 16]
```diff
@@ -5841,6 +5841,7 @@ impl<Signer: EcdsaChannelSigner> ChannelMonitorImpl<Signer> {
 								on_to_local_output_csv: None,
 							},
 						});
+						self.counterparty_fulfilled_htlcs.insert(SentHTLCId::from_source(&source), payment_preimage);
```
Comment (wpaulino) [inline lightning/src/chain/channelmonitor.rs:5958]: IIUC this change will now result in users seeing duplicate `PaymentSent/Forwarded` events after every reload once the channel closes?

File: lightning/src/chain/channelmonitor.rs [hunk 17]
```diff
@@ -3034,6 +3034,125 @@ impl<Signer: EcdsaChannelSigner> ChannelMonitor<Signer> {
...
+					if let &Some(ref source) = b {
+						Some((a, Some(&**source)))
+					} else {
+						None
+					}
+				})
+			);
+		} else if confirmed_txid == funding.current_holder_commitment_tx.trust().txid() {
+			walk_htlcs!(true, holder_commitment_htlcs!(us, CURRENT_WITH_SOURCES));
+		} else if let Some(prev_commitment_tx) = &funding.prev_holder_commitment_tx {
```
Comment (wpaulino) [inline lightning/src/chain/channelmonitor.rs:3136]: We could map the option to the `confirmed_txid == prev_commitment_tx.trust().txid()` bool and avoid the inner else branch, but maybe you find this less readable

File: lightning/src/chain/channelmonitor.rs [hunk 18]
```diff
@@ -3034,6 +3034,125 @@ impl<Signer: EcdsaChannelSigner> ChannelMonitor<Signer> {
...
+							// HTLC has been failed.
+							res.insert(source.clone(), candidate_htlc.payment_hash);
+						}
+					}
+				};
+
+				// We walk the set of HTLCs in the unrevoked counterparty commitment transactions (see
+				// `fail_unbroadcast_htlcs` for a description of why).
+				if let Some(ref txid) = us.funding.current_counterparty_commitment_txid {
+					if let Some(htlcs) = us.funding.counterparty_claimable_outpoints.get(txid) {
```
Comment (wpaulino) [inline lightning/src/chain/channelmonitor.rs]: We should `expect` this, no?

File: lightning/src/chain/channelmonitor.rs [hunk 19]
```diff
@@ -3034,6 +3034,125 @@ impl<Signer: EcdsaChannelSigner> ChannelMonitor<Signer> {
...
+		});
+
+		let confirmed_txid = if let Some(txid) = confirmed_txid {
+			txid
+		} else {
+			return res;
+		};
+
+		macro_rules! walk_htlcs {
+			($holder_commitment: expr, $htlc_iter: expr) => {
```
Comment (wpaulino) [inline lightning/src/chain/channelmonitor.rs]: `$holder_commitment` looks unused

File: lightning/src/ln/channelmanager.rs (hunk 14 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs]: We're required to have a peer state around for any channelmonitor we have (to store the latest-update-id of the monitor), so its not a bug, but I agree it should be cleaned up.

File: lightning/src/chain/channelmonitor.rs (hunk 16 above)
Comment (TheBlueMatt) [reply lightning/src/chain/channelmonitor.rs:5958]: Yes, that is fixed by #3984

File: lightning/src/chain/channelmonitor.rs (hunk 17 above)
Comment (TheBlueMatt) [reply lightning/src/chain/channelmonitor.rs:3136]: Its kinda hard to do without rustfmt vomiting all over the place :(

File: lightning/src/chain/channelmonitor.rs (hunk 18 above)
Comment (TheBlueMatt) [reply lightning/src/chain/channelmonitor.rs]: We can, yea.

File: lightning/src/ln/channelmanager.rs (hunk 14 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs]: Note that this was copied from the later block.

**Comment by valentinewallace** [comment]
Some tests are failing

**Comment by TheBlueMatt** [comment]
Huh, must be more rebase to hit that.

**Comment by TheBlueMatt** [comment]
Squashed without further changes.

**Comment by TheBlueMatt** [comment]
Backported in #4143
